src/main/java/io/cdap/plugin/ariba/source/AribaServices.java

```diff
@@ -207,26 +336,116 @@ private String getBase64EncodedValue() {
   /**
    * Prepares output schema based on the provided plugin config parameters.
    */
-  public Schema buildOutputSchema(String accessToken) throws IOException {
+  public Schema buildOutputSchema(String accessToken) throws IOException, AribaException, InterruptedException {
     LOG.trace("Initiating Metadata Call To Ariba");
-    AribaColumnMetadata aribaColumnMetadataList = getMetadata(accessToken);
+    List<AribaColumnMetadata> aribaColumnMetadataList = getMetadata(accessToken);
     AribaSchemaGenerator schemaGenerator = new AribaSchemaGenerator(aribaColumnMetadataList);
     return schemaGenerator.buildSchema();
   }
 
   /**
    * Fetch metadata from Ariba.
+   *
+   * @param accessToken {@code Access Token}
    */
-  private AribaColumnMetadata getMetadata(String accessToken) throws IOException {
-    AribaResponseContainer responseContainer = fetchAribaResponse(generateMetadataURL(accessToken), accessToken);
-    try (InputStream responseStream = responseContainer.getResponseBody()) {
-      ObjectMapper objectMapper = new ObjectMapper();
-      JsonNode jsonNode = objectMapper.readTree(responseStream);
+  public List<AribaColumnMetadata> getMetadata(String accessToken) throws
+    IOException, AribaException, InterruptedException {
+    HttpUrl.Builder templateBuilder = metadataTemplateBuilder(false, null);
+    AribaResponseContainer responseContainer = fetchAribaResponse(templateBuilder.build().url(), accessToken);
+    InputStream responseStream = responseContainer.getResponseBody();
+    JsonNode jsonNode = objectMapper.readTree(responseStream);
+    if (responseContainer.getHttpStatusCode() == HttpURLConnection.HTTP_OK) {
+
+      AribaMetaResponse res = gson.fromJson(jsonNode.toString(), AribaMetaResponse.class);
+      // Simple Fields
+      List<AribaColumnMetadata> columnDetails =
+        generateColumnValues(Properties.getNonObjectFields(res.getProperties()));
+      // Object Fields
+      columnDetails.addAll(getObjectFields(Properties.getObjectFields(res.getProperties())));
+      // Array Fields
+      columnDetails.addAll(getArrayFields(Properties.getArrayFields(res.getProperties())));
+      return columnDetails;
+    }
+    String errMsg = jsonNode.get(ResourceConstants.MESSAGE).asText() != null
+      ? jsonNode.get(ResourceConstants.MESSAGE).asText() :
+      ResourceConstants.ERR_NOT_FOUND.getMsgForKey();
+    throw new AribaException(errMsg, responseContainer.getHttpStatusCode());
+  }
+
+  /**
+   * @param aribaPluginConfig ariba plugin config
+   * @return JsonNode
+   */
+  public JsonNode createJob(AribaPluginConfig aribaPluginConfig, String pageToken) throws
+    AribaException, IOException, InterruptedException {
+    Request req = buildJobRequest(jobBuilder(pageToken).build().url(), aribaPluginConfig);
+    Response response;
+    do {
+      response = executeRequest(req);
+      if (response.code() == ResourceConstants.API_LIMIT_EXCEED) {
```

Comment:
Can we quickly discuss this on call?
I remember we had a discussion on this with you and Albert both and we finalised that only for day limit we should throw exception. Only change Albert suggested to make it simple is to remove the logic of writing pageToken on gcs location.
And I have added check for MAX_RETRIES so that it doesn’t look like an infinite loop.